### PR TITLE
Close the pet persistence durability boundaries as accepted limits

### LINKED ISSUE #162: Pets: close the remaining persistence durability boundaries
# NEW OVERRIDING RULE

This got over-engineered in scope.  We want a minimal, fast implementation.  Losing pet data is not a big deal.

## Problem

The pet persistence refactor closed gameplay parity (Phase 1, evidence in `docs/testing/pet-phase1-*-2026-09-09.txt`) and the follower-category, expiry, and schema-identity work (#118, #119, #120). The remaining durability boundaries from the Phase 2 plan are still open. `docs/systems/SAVE_SYSTEMS_BREAKDOWN.md` (Persistent Followers) currently states that post-publication callback reconciliation, expiry gear handling, and uncertain-commit reconciliation remain open.

## Remaining work

1. **Callback and transaction edge cases**
   - Define and test equipment and mobile callback behavior while a pet is prepared outside a room.
   - Reconcile extraction or mutation during publication without losing the saved pet or its items.
   - Reconcile uncertain commit outcomes for active saves, keeper storage, and keeper reclaim.
   - Document the boundary between SQL pet state, pfiles, and live world objects instead of claiming cross-store crash atomicity.
2. **Legacy persistence compatibility**
   - Verify stable identity, owner binding, rename and name-reuse isolation, nested object ordering, equipment association, and schema upgrades against existing saved data.
   - The earlier plan claimed some pet container rows carry custom weights inflated before a container-weight normalization fix. That claim is unverified (no changelog or source evidence was found on 2026-09-12); confirm it with real saved data before designing a repair, or drop it.
   - Remove a superseded pet path only after tracing that it has no remaining callers.
3. **Tests**: extend the existing production-linked suites (`test_pet_policy.c`, `test_pet_persistence.c`, `test_database_persistence.c`, `test_gameplay_e2e.c`, and the relevant spell, class, crafting, combat, mount, and object suites) to cover success, denial, resource retention or consumption, expiry, death, separation, malformed data, retry, duplicate prevention, callback extraction, and rollback. Use isolated MariaDB for database bodies; do not add a parallel pet test harness.
4. **Final integration acceptance**: exercise summoner, necromancer, divine, mount, construct, purchased-pet, and source-item gameplay in the development MUD through `autorun.sh` on port 4100, including equipment and nested containers, owner death and reclaim, logout and relogin, restart, actual copyover, finite expiry, reduced-capacity restore, and failed-operation recovery. Run `LUMINARI_TEST_MYSQL_ENABLE=1 make -j$(nproc) test`, `make install`, the CMake build, and focused memory checks on changed parsing, lifetime, object, and callback paths.
5. **Documentation and help**: update command and feature help in both `lib/text/help/help.hlp` and the development help database (with tracked SQL components where publication is required), and update `SAVE_SYSTEMS_BREAKDOWN.md` to describe final behavior and known durability limits.

## Completion criteria

- Each boundary above has either a tested implementation or a documented, explicit limitation in `docs/systems/SAVE_SYSTEMS_BREAKDOWN.md`.
- The copyover and restart acceptance run is recorded with the tested revision.
- No "remains open" pointer to a temporary plan document is left in permanent documentation.

# NEW OVERRIDING RULE

This got over-engineered in scope.  We want a minimal, fast implementation.  Losing pet data is not a big deal.

### PR DESCRIPTION
**Pull request by moshehbenavraham**
Closes #162 under its overriding rule (minimal pass; losing pet data is acceptable).

## What changed

- `docs/systems/SAVE_SYSTEMS_BREAKDOWN.md`: the post-publication callback, expiry-gear, uncertain-commit, and legacy-data boundaries are now recorded as explicit accepted limits, traced from the current `save_char_pets()`, keeper store, and keeper reclaim code. The "remains open" pointers to the issue are gone.
- `docs/testing/pet-copyover-restart-acceptance-2026-09-12.txt`: records the copyover and full-restart acceptance run with the tested revision.
- `src/players.c`: the live run exposed a defect. Restoring a purchased mercenary forced `MOB_MERCENARY` from its prototype special, so a pet listed as General before a copyover came back as Mercenary in a different capacity slot. Restore now carries only the hit-point roll marker; a flag that was set on the live mobile still round-trips through the runtime state's kept-flag line (the existing round-trip test covers that path).

## Verification

- `LUMINARI_TEST_MYSQL_ENABLE=1 make -j$(nproc) test`, with the database tests rerun against the development database: `OK (1424 tests)`.
- `make install` clean, no warnings, no root `luminari` artifact.
- Live on port 4100: purchase, gear, copyover (same row id, bag retained), second purchase stays General across copyover and relogin, `shutdown reboot` restart restores both pets with gear, cleanup back to 0 pets.

## Not done (by the minimal rule)

No new code for cross-store reconciliation, expiry gear return, or legacy data repair; no help changes (no command behavior changed); CMake build and memory checks not rerun for this three-line change.

## Summary by CodeRabbit

* **Bug Fixes**
  * Restored hired followers now retain their correct category and status.
  * Pet restoration and persistence behavior is more reliable across copyovers and restarts.
  * Save failures are handled consistently, with documented limits for uncertain outcomes.

* **Documentation**
  * Updated pet persistence documentation to clarify durability limits and recovery behavior.
  * Added an acceptance record covering pet purchase, copyover, restart, and cleanup scenarios.

### CODE REVIEW

File: docs/systems/SAVE_SYSTEMS_BREAKDOWN.md [hunk 1]
```diff
@@ -116,9 +116,12 @@ every follower row before starting the transaction, then replaces active pet and
...
 including across a rename. SQL player IDs are not interchangeable with pfile IDs.
-A known query failure rolls back the replacement. An uncertain COMMIT outcome
-still needs reconciliation; a transaction does not make pfiles, world items,
-and SQL jointly crash-atomic.
+A known query failure rolls back the replacement. A COMMIT that reports an
+error is treated as a failure even though the server may already have applied
+it: the live pets stay in play and the next snapshot replaces the active rows
+again, so the worst case is a stale active row until then. A transaction does
+not make pfiles, world items, and SQL jointly crash-atomic; pet data is not
+treated as critical, and no cross-store reconciliation is attempted.
```
[the remark this answers is not in the document]
Comment (moshehbenavraham) [reply docs/systems/SAVE_SYSTEMS_BREAKDOWN.md]: Addressed in f18aca377: the passage now states that both failed-COMMIT paths issue a ROLLBACK with an uncertain effect, and that a keeper reclaim whose publication fails may leave the row active if the separate recovery transaction also fails, with the owner's next login restoring it from there.

File: docs/systems/SAVE_SYSTEMS_BREAKDOWN.md [hunk 2]
```diff
@@ -196,8 +199,16 @@ are moved to `PET_STATE_STORED` in one transaction before anything is published,
...
+the prototype. A restored hireling keeps the prototype's category; the runtime
+state carries only its one-time hit-point roll marker plus any flag that was
+set on the live mobile, so a purchased mercenary counts against the same slot
+before and after a restore. Placement and mobile load triggers that extract or
+detach the pet during publication discard the roomless copy and leave the row
+saved. Once published, the pet is an ordinary live NPC: anything that happens
+to it afterwards is captured by the next snapshot or lost with it, and no
+further reconciliation is performed. Legacy rows with a NULL runtime state
+retain the compatibility load path; older saved data is not repaired or
+migrated beyond the schema migrations. Custom names use the existing name/short/long text fields.
```
[the remark this answers is not in the document]
Comment (moshehbenavraham) [reply docs/systems/SAVE_SYSTEMS_BREAKDOWN.md:217]: Addressed in f18aca377: the passage now states that both failed-COMMIT paths issue a ROLLBACK with an uncertain effect, and that a keeper reclaim whose publication fails may leave the row active if the separate recovery transaction also fails, with the owner's next login restoring it from there.

**Comment by moshehbenavraham** [comment]
Review addressed and squashed to a single commit (531b87bee). Branch is on top of current master (no rebase needed).

Both CodeRabbit findings were verified against the code and fixed in SAVE_SYSTEMS_BREAKDOWN.md: every failed-COMMIT path issues a ROLLBACK with an uncertain effect, and a keeper reclaim whose publication fails can leave its row active if the recovery transaction also fails (the owner's next login restores it). apply_pet_runtime_state() now carries a description comment.

Local runs of the CI jobs on this commit:

- Source hygiene: 1613 files clean. Build manifest parity: OK. SQL interpolation: within baseline.
- clang-format (pre-commit, all files): passed. clang-tidy CI subset: same pre-existing notices, none on changed lines.
- Gitleaks on the tracked tree: no leaks.
- World-data tool tests, constants sync, docs check: OK.
- Autotools make test with database tests: OK (1424 tests); make install clean, no warnings.
- CMake ci-gcc preset with LUMINARI_WERROR=ON: built with zero warnings; ctest 26/26 passed, production cutest OK (1424 tests).
- Live server: copyover onto the installed build passed; earlier full restart run is in the acceptance record.

Not run locally: CodeQL and Dependency Review (GitHub-hosted), the clean-archive job and ci-clang matrix leg (no source files added or removed), and the integration schema dry-run (needs database admin rights this checkout does not have; database tests ran against the development database because a separate isolated schema cannot be created here).

Ready to merge; squash merge is fine.
